src/ai-function/types.ts

```diff
@@ -0,0 +1,68 @@
+import { type SetOptional } from 'type-fest';
+import { type z } from 'zod';
+
+import { type Model, type Msg } from '../model/types.js';
+
+/**
+ * A runner that iteratively calls the model and handles function calls.
+ */
+export type Runner<Content = string> = (
+  params: string | Runner.Params,
+  context?: Model.Ctx
+) => Promise<Runner.Response<Content>>;
+
+export namespace Runner {
+  /** Parameters to execute a runner */
+  export type Params = SetOptional<Model.Chat.Run & Model.Chat.Config, 'model'>;
+
+  export type ModelParams = Partial<
+    Omit<Model.Chat.Run & Model.Chat.Config, 'messages' | 'functions' | 'tools'>
+  >;
+
+  /** Response from executing a runner */
+  export type Response<Content = string> =
+    | {
+        status: 'success';
+        messages: Msg[];
+        content: Content;
+      }
+    | {
+        status: 'error';
+        messages: Msg[];
+        error: Error;
+      };
+
+  /** Controls use of functions or tool_calls from OpenAI API */
+  export type Mode = 'tools' | 'functions';
+}
+
+/**
+ * A function used to extract data using OpenAI function calling.
+ */
+export type ExtractFunction<Schema extends z.ZodObject<any>> = (
+  params: string | Runner.Params,
+  context?: Model.Ctx
+) => Promise<z.infer<Schema>>;
+
+export interface AIFunctionSpec {
+  name: string;
+  description?: string;
+  parameters: Record<string, unknown>;
+}
+
+/**
+ * A function meant to be used with OpenAI function calling.
+ */
+export interface AIFunction<
+  Schema extends z.ZodObject<any> = z.ZodObject<any>,
+  Return = any,
```

Comment:
in the future, we should ideally restrict the return to be jsonifiable; let's ignore for now

Reply:
Or at least a nominal `type AnyJson = number | string | boolean | object | etc` so there is a small amount of direction there for documentation purposes.